### PR TITLE
Add developer side walkthrough

### LINKED ISSUE #193: Document how to make simple, developer-side model edits
This issue is to list developer-side edits that are relatively simple to make, and to request a vignette documenting how they should be made within the package code.

### PR DESCRIPTION
**Pull request by pratikunterwegs**
This PR aims to fix #193 by adding a developer side walkthrough vignette that lays out how to make some quick edits to the ODE models.

### CODE REVIEW

File: vignettes/developer-walkthrough.Rmd [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+This vignette is structured around use cases of making package modifications given the existing package architecture.
+
+Consequently this vignette does not cover major changes that would require a substantial rewrite of the package, such as switching the C++ solver used for ODE models, or switching to using _cpp11_ rather than _Rcpp_.
+
+This vignette also does not tackle questions around epidemic modelling choices - such as which compartmental transitions should be allowed; we strongly recommend the involvement of a specialist in such cases.
+
+::: {.alert .alert-info}
+**Making substantial changes to a model in epidemics?**
+Have you modified an existing model to cover new use cases or disease types? We welcome contributions of new model structures. Examples include models targeting vector borne diseases, or diseases with a sylvatic cycle. Please get in touch via a GitHub issue or Discussion!
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:39]:
```suggestion
Have you modified an existing model to cover new use cases or disease types? We welcome contributions of new model structures. Examples include models targeting vector borne diseases, or diseases with a sylvatic cycle. Please get in touch via a GitHub [Issue](https://github.com/epiverse-trace/epidemics/issues)!
```

There's no discussion board, right?

File: vignettes/developer-walkthrough.Rmd [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+1. First, navigate to the end of each model `FunctionObject` defined in `inst/include/model_*.h` and edit the ODEs found there (see examples below). Users can hard-code the parameter values in these files for quick edits, but we recommend taking the next few steps to allow changing these values from R.
+
+2. Next, modify the C++ function arguments in `src/model_*.cpp` so that any newly added rate parameters can be passed from R to C++. Typically, single parameters should be passed by reference as `const double`.
+
+  - If you seek to use time-dependence and rate interventions with the newly added parameters should follow the steps in the section above.
+
+3. Finally, modify the R code in `R/model_*.R` so that the R function accepts the new parameter as an appropriately checked argument.
+
+4. When adding parameters such as birth and death rates, make sure that this is accounted for in any unit tests (see especially the current expectation in most models that there is a constant population size).
+
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:94]: I'd find the vignette easier to follow if there was a first section "Adding/removing parameters" with this content, then making them time-dependent, then changing flows.

File: vignettes/developer-walkthrough.Rmd [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+**Note that** existing compartmental flows are typically controlled by infection parameters, which can be set to zero to disallow specific transitions. For example, to simulate a scenario in which hospitalisation is not available, the Vacamole model can be run with the argument `hospitalisation_rate = 0.0`.
+
+Please check whether setting an infection parameter to zero better suits your use case.
+:::
+
+This section focuses on adding novel inflows and outflows via some illustrative examples.
+
+1. First, navigate to the end of each model `FunctionObject` defined in `inst/include/model_*.h` and edit the ODEs found there (see examples below). Users can hard-code the parameter values in these files for quick edits, but we recommend taking the next few steps to allow changing these values from R.
+
+2. Next, modify the C++ function arguments in `src/model_*.cpp` so that any newly added rate parameters can be passed from R to C++. Typically, single parameters should be passed by reference as `const double`.
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:87]: How would one add parameters of dimension >1 (e.g. age-dependent)?

I later found it below. Perhaps link to that section later?

File: vignettes/developer-walkthrough.Rmd [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+Consequently this vignette does not cover major changes that would require a substantial rewrite of the package, such as switching the C++ solver used for ODE models, or switching to using _cpp11_ rather than _Rcpp_.
+
+This vignette also does not tackle questions around epidemic modelling choices - such as which compartmental transitions should be allowed; we strongly recommend the involvement of a specialist in such cases.
+
+::: {.alert .alert-info}
+**Making substantial changes to a model in epidemics?**
+Have you modified an existing model to cover new use cases or disease types? We welcome contributions of new model structures. Examples include models targeting vector borne diseases, or diseases with a sylvatic cycle. Please get in touch via a GitHub issue or Discussion!
+:::
+
+## Modifying which parameters can be time dependent
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:42]: Before that can we have a small section "where to find the model" describing the file/directory structure (as models are spread across multiple directories)?

File: vignettes/developer-walkthrough.Rmd [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+4. When adding parameters such as birth and death rates, make sure that this is accounted for in any unit tests (see especially the current expectation in most models that there is a constant population size).
+
+### Modelling births to $S$ and deaths from all compartments
+
+Define birth rate and background mortality rate parameters (denoted $b$ and $d$ respectively), and modify the compartmental transitions as follows. This example shows only the equations for the susceptible compartment.
+
+```c++
+// Add births to S, where b is the birth rate, and N is the population size
+dxdt.col(0) = B -sToE - sToV;  // B -β*S*contacts*I - ν
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:101]: Can you show what the line looked like before and clarify that this shows the ODE for the number of susceptibles (just so people don't think this is a new line to be added). I.e. in the previous section (with the header one level higher) could say: "In the following, we will consider a number of modifications to the susceptible compartment, which we assume to be given by...".

File: vignettes/developer-walkthrough.Rmd [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+  - If you seek to use time-dependence and rate interventions with the newly added parameters should follow the steps in the section above.
+
+3. Finally, modify the R code in `R/model_*.R` so that the R function accepts the new parameter as an appropriately checked argument.
+
+4. When adding parameters such as birth and death rates, make sure that this is accounted for in any unit tests (see especially the current expectation in most models that there is a constant population size).
+
+### Modelling births to $S$ and deaths from all compartments
+
+Define birth rate and background mortality rate parameters (denoted $b$ and $d$ respectively), and modify the compartmental transitions as follows. This example shows only the equations for the susceptible compartment.
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:97]:
```suggestion
Define birth rate and background mortality rate parameters (denoted $B$ and $D$ respectively), and modify the compartmental transitions as follows. This example shows only the equations for the susceptible compartment.
```

File: vignettes/developer-walkthrough.Rmd [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+  
+  - To simulate births as the only inflow in an age-stratified model, `B` must have the same length as `N`, with only the first element having a non-zero value. `B` is still required to be a vector for compatibility with other vector/array elements. Setting other values of `B` to be non-zero positive values can be used to model immigration or other long-term processes.
+
+    ```c++
+    // where `x` is the current state matrix
+    // get current population size
+    // recall rows are demography groups, columns are compartments
+    const Eigen::ArrayXd population_size = x.rowwise().sum();
+
+    // define a birth rate
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:117]:
```suggestion
    // define a per-capita birth rate
```

File: vignettes/developer-walkthrough.Rmd [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    const double b = 1e-3;
+
+    // calculate number of births
+    double births_ = (b * population_size.sum());
+    
+    // create the vector `B`
+    Eigen::ArrayXd births(4);  // hardcoded for four age groups
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:124]: why not call it `B` as in the example above? Or is this something else?

File: vignettes/developer-walkthrough.Rmd [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+    ```c++
+    // where `x` is the current state matrix
+    // get current population size
+    // recall rows are demography groups, columns are compartments
+    const Eigen::ArrayXd population_size = x.rowwise().sum();
+
+    // define a birth rate
+    const double b = 1e-3;
+
+    // calculate number of births
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:120]:
```suggestion
    // calculate total birth rate
```
just sticking with the terminology from above

File: vignettes/developer-walkthrough.Rmd [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+
+    ```c++
+    // a uniform value for the rate of immigration in all age groups
+    const double immigration_rate = SOME_VALUE;
+
+    // calculate immigration as a proportion of each age group
+    const Eigen::ArrayXd immigration = immigration_rate * population_size;
+
+    // sum births and immigration
+    const Eigen::ArrayXd all_inflows = births + immigration;
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:141]: and then replace `B` with `all_inflows` in line 101?

File: vignettes/developer-walkthrough.Rmd [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+### Modelling waning immunity
+
+Similar to previous examples, define a uniform waning rate for the immunity of recovered individuals and modify the appropriate compartmental transitions.
+
+```c++
+// define an age-specific waning rate
+const double waning_rate = SOME_VALUE;
+
+// include this new R -> S transition
+// note x.col(3).array() is the vector of age-specific recovereds
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:170]: why is this 3? Is this defined somewhere?

File: vignettes/developer-walkthrough.Rmd [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+Adding epidemiological compartments is similar to modifying compartmental transitions, but requires a few extra steps.
+
+1. Add and modify model ODEs as appropriate in `inst/include/model_*.h` (see sections above).
+
+2. Modify the `compartments` vector in `R/model_*.R` to include the name of the new compartment. This is used both in checking that the input `<population>` has the correct number of compartments in its initial state matrix, as well as for formatting the output data.
+
+## Changing vaccination rates over time
+
+_epidemics_ treats vaccination as special process that requires a composable element in the form of an S3 class, `<vaccination>`, which allows defining time-limited, group-specific vaccination rates for multiple vaccination doses if necessary.
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:204]: And I can't make these time-limited rates different at different times, right?

**Review by sbfnk — CHANGES_REQUESTED** [review]
Thanks, I think this will be really useful for people wanting to make changes to the model(s). I left a few comments where I stumbled across things.

Can you add a section "how to add a new model"? I.e. should I start with copying the default model and then take it from there?

File: vignettes/developer-walkthrough.Rmd [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+Vaccination is only strictly allowed in the ODE models 'default' and 'Vacamole', and these are covered here. Begin by dismantling the model code to remove any functionality related to the `<vaccination>` class.
```
Comment (sbfnk) [inline vignettes/developer-walkthrough.Rmd:206]: Should we just define a default model with and one without vaccination?

**Comment by pratikunterwegs** [comment]
Thanks @sbfnk - agree with all. Will wait perhaps a week or two in case more feedback comes in.

File: vignettes/developer-walkthrough.Rmd (hunk 4 above)
Comment (adamkucharski) [reply vignettes/developer-walkthrough.Rmd:42]: Agree – having this at the start to orientate contributors would be very helpful, i.e. how `R/model_default.R`, `R/check_args_default.R`, `src/model_default.cpp`,  `inst/include/model_default.h` link together with the `model_default()` function call.

File: vignettes/developer-walkthrough.Rmd (hunk 11 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd:170]: It's hardcoded as the column holding the recovered; this has been added to the model header file.

File: vignettes/developer-walkthrough.Rmd (hunk 12 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd:204]: That's right, but this section shows how to remove this functionality and implement a uniform vaccination rate that could be targeted with the time-dependence functionality to make the rate different at different times.

File: vignettes/developer-walkthrough.Rmd (hunk 13 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd:206]: I would hesitate to do that as one can get a no-vaccination model by simply not passing a `<vaccination>` to the model. Adding another model with such a small difference would add to the package code without substantial gain in functionality.

File: vignettes/developer-walkthrough.Rmd (hunk 10 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd:141]: Yes, added.

File: vignettes/developer-walkthrough.Rmd (hunk 8 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd:124]: Agreed, but then thought I should have clearly named variables even in this dummy example.

File: vignettes/developer-walkthrough.Rmd (hunk 1 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd]: Added in https://github.com/epiverse-trace/epidemics/pull/210/commits/d0646b6d71f2e1f5624db04db670c0b48dbe575f

File: vignettes/developer-walkthrough.Rmd (hunk 2 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd:122]: Added in https://github.com/epiverse-trace/epidemics/pull/210/commits/8085ddef0037c295895ad630cd8af8fed8cae5a2. I've moved the sections around a bit, hopefully more understandable now

File: vignettes/developer-walkthrough.Rmd (hunk 4 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd]: Added a directory tree with comments that hopefully explains this in https://github.com/epiverse-trace/epidemics/pull/210/commits/4df85c0b353332bffbbad56fb17108303c3c477f

File: vignettes/developer-walkthrough.Rmd (hunk 5 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd]: Added in https://github.com/epiverse-trace/epidemics/pull/210/commits/440ad2e49dbeb9f63fe1477fbc876f0924a12d0d

File: vignettes/developer-walkthrough.Rmd (hunk 6 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd]: Changed this a bit in https://github.com/epiverse-trace/epidemics/pull/210/commits/440ad2e49dbeb9f63fe1477fbc876f0924a12d0d, hope it makes sense

File: vignettes/developer-walkthrough.Rmd (hunk 7 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd]: Fixed in https://github.com/epiverse-trace/epidemics/pull/210/commits/440ad2e49dbeb9f63fe1477fbc876f0924a12d0d

File: vignettes/developer-walkthrough.Rmd (hunk 9 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd]: Fixed in https://github.com/epiverse-trace/epidemics/pull/210/commits/440ad2e49dbeb9f63fe1477fbc876f0924a12d0d I think - but do say if terminology is wrong

File: vignettes/developer-walkthrough.Rmd (hunk 3 above)
Comment (pratikunterwegs) [reply vignettes/developer-walkthrough.Rmd:115]: Added links to this section

**Comment by pratikunterwegs** [comment]
Thanks @sbfnk and @adamkucharski for taking a look at this. I've pushed changes that show the minimal package structure, and a section on adding and removing parameters. Hope this makes it easier to read. I'm aiming to merge this by the end of the week.

**Comment by alxsrobert** [comment]
Hi Pratik, Adam asked me last week to have a look at the developer vignette for the perspective of someone who is not very familiar with the package (so I've also looked at the "design decision" vignette to understand the structure of epidemics). I've written a few notes. Overall I thought the vignette was clear and well-designed, but I thought of some clarifications:
- I think that the  "guide to package structure" section in the developer walkthrough does not make sense to someone who is not already quite familiar with the package:
  - Some explanations of the files do not really make sense to me ("functions to construct cross-checking fns", "MODEL FUNC WRAPPING ODE SYSTEM; LEVEL 2")
  - The levels are not yet defined and therefore appear confusing.
  - I would suggest removing the mentions of scripts that are not mentioned in this vignette (so have a "simplified" structure, and maybe have the full structure at the end of the vignette, where there can be a slightly longer description of each file ). And introduce what the levels correspond to before the package structure.
- Similarly, in the "design decisions"  vignette, I would put figure 1 and 2 (so the "package architecture" section) after the definitions of some terms and broad concepts:
  - Figure 2 is quite overwhelming, so I didn't really know what I was meant to look for and take from it. I was also wondering what terms like "composable elements" from Figure 1 meant, only to have it defined at the next section, it would be better if it was defined before the figure (or right after it). 
  - I would suggest adding a few sentences explaining what the figure is for / what we should look at. Maybe also splitting Figure 2 into multiple figures?
- I think the section "adding or removing parameter" could do with some example code for each subsection. For instance, "Modify the compartmental transitions in the FunctionObject ODE model inst/include/model_*.h appropriately to include the mortality rate" seems unclear to me, does that just refer to adding / removing an extra argument? Having the description without the example makes it sound complicated, while if I understand well each step requires little re-coding (for a simple example).
- I think the sections with examples are good and clear. It would be good to remind the reader in what script each snippet of code is added (I think they are mentioned in the "Modifying compartmental flow..." section heading, but it would be clearer if it was repeated here).
- I'm don't think that every single section needs a specific example, but in my opinion the vignette should start with walk-through examples (like in "modelling birth"/ "modelling waning immunity"), so the reader would start with simple code, and become more familiar with the structure, before sections that just feature explanations.

I hope this is helpful, let me know if anything is unclear!

**Comment by pratikunterwegs** [comment]
Thanks @alxsrobert - all helpful suggestions that I'll incorporate as edits to this document.

**Pull request merged**